Add download functionnality

vincent-boulin opened the pull request:
Hello,

I'm user of your grunt plugin grunt-ftp, and for some projects we need to put and get some files in an FTP Server.
Like i don't find some grunt plugin which do the both, i add this functionnality into your plugin.

Let me know if what i've done is great or if it's need to change some other things.

I don't have think of comptability with older version, so maybe we will need to rename ftpPut task to its older name.

PS : Sorry my english, i'm french.

Regards,

vincent-boulin committed:
Add ftpGet task to download

download file from a ftp server

vincent-boulin committed:
Rename task and update readme.md

Rename ftp task to ftpPut
Update readme.md
Update version in package.json

sindresorhus commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "grunt-ftp",
-  "version": "1.0.2",
+  "version": "1.1.0",
```
don't bump

sindresorhus commented on tasks/ftp.js:
```diff
@@ -56,4 +56,58 @@ module.exports = function (grunt) {
+			// have to create a new connection for each file otherwise they conflict
+			var ftp = new JSFtp(options);
+
+			if(!grunt.file.exists(el.dest)) {
```
No need to `if` check it, just use `grunt.file.mkdir` directly.

sindresorhus commented on readme.md:
```diff
@@ -35,6 +35,28 @@ grunt.initConfig({
+			files: {
+				'public_html': 'src/file.txt'
+			}
+		}
```
this needs to be updated

sindresorhus commented on readme.md:
```diff
@@ -35,6 +35,28 @@ grunt.initConfig({
+	}
+});
+
+grunt.registerTask('default', ['ftp']);
```
same

sindresorhus commented on readme.md:
```diff
@@ -58,6 +80,9 @@ Default: `'anonymous'`
 
+## Notes
+
+For downloading task, if dest path is a folder that doesn't exists, this will be create
```
not needed

sindresorhus commented on tasks/ftp.js:
```diff
@@ -56,4 +56,58 @@ module.exports = function (grunt) {
+				// if dest is a directory, have to create a file with source filename
+				var filename = path.basename(el.src[0]);
+				finalLocalPath = path.join(el.dest, '/', filename);
+			}
```
```js

vincent-boulin committed:
Take account of remarks

Don't need to test existence of dest folder, grunt.file.mkdir do it.
Update of package.json.
Update of readme.md for forgets

sk4ry commented:
Hello,

I've considered your remarks.
Let me know, if you want me to update the ftpPut task for the backward compatibility.

Thank you.

Regards,

sindresorhus commented:
> Let me know, if you want me to update the ftpPut task for the backward compatibility.

No need. I'll just alias it back after merging this.

sindresorhus commented on tasks/ftp.js:
```diff
@@ -56,4 +56,55 @@ module.exports = function (grunt) {
+			grunt.file.mkdir(path.dirname(el.dest));
+
+			var finalLocalPath = el.dest;
+			if(grunt.file.isDir(el.dest)) {
```
`if (grunt.file.isDir(el.dest)) {`

sk4ry replied:
This test is used for this config example :
```js
files: {
	'fixtureGet/fixture2.txt': 'ftp/fixture/fixture.txt',
	'<%= dirs.local %>/fixtureGet': 'ftp/fixture/fixture.txt'
}
```

If the dest param is a directory, we create a new file on path : "dest_dir/src_filename".
May be there is a better way to do this ?

sindresorhus replied:
@sk4ry I was only commenting on the code style. Notice the difference.

sk4ry replied:
Oh ok, i don't notice it first time.
I'll commit this so.

sindresorhus commented on test.js:
```diff
@@ -5,3 +5,12 @@ var grunt = require('grunt');
 	assert(grunt.file.exists('ftp/fixture/fixture.txt'));
 });
+
+it('should create a new folder for retrieve files', function() {
```
`function () {`

sk4ry replied:
Is the test of folder creation useless ?

sindresorhus replied:
Code style.

vincent-boulin committed:
Code style

Code style remarks

sk4ry commented:
Normally it's good now.

Just a question for my knowledge, why do you say me not to bump the version ?

sindresorhus commented:
> Just a question for my knowledge, why do you say me not to bump the version ?

Because it's not up to you what version I will release it as. I might want to do more changes before a release. I might have a release script which bumps it automagically (I actually do). Just don't bump other peoples versions ;)

sk4ry commented:
Ok, effectively it makes sense :). Thank you for the tip.
